lib/Sema/CSDiagnostics.h

```diff
@@ -777,6 +777,28 @@ class AutoClosureForwardingFailure final : public FailureDiagnostic {
   bool diagnoseAsError() override;
 };
 
+/// Diagnose invalid pointer conversions for an autoclosure result type.
+///
+/// \code
+/// func foo(_ x: @autoclosure () -> UnsafePointer<Int>) {}
+///
+/// var i = 0
+/// foo(&i) // Invalid conversion to UnsafePointer
+/// \endcode
+class AutoClosurePointerConversionFailure final : public FailureDiagnostic {
```

Comment:
I think we can consider this a contextual failure so parent should be `ContextualFailure` here which also has storage for both types.